torch/lib/c10d/TCPStore.cpp

```diff
@@ -656,12 +657,32 @@ class TCPServer {
   TCPSocket socket_;
   PortType port_;
   std::unique_ptr<TCPStoreMasterDaemon> daemon_;
+
+  // We store weak references to all TCPServers for which the caller requested
+  // multi-tenancy.
+  static std::unordered_map<PortType, std::weak_ptr<TCPServer>> cachedServers_;
 };
 
+std::unordered_map<PortType, std::weak_ptr<TCPServer>>
+    TCPServer::cachedServers_{};
+
 std::shared_ptr<TCPServer> TCPServer::start(const TCPStoreOptions& opts) {
-  if (opts.multiTenant) {
-    LOG(WARNING)
-        << "The multi-tenant feature of TCPStore is not implemented yet.";
+  std::shared_ptr<TCPServer> server{};
+
+  // If the caller is okay with a multi-tenant store, first check if we already
+  // have a TCPServer running on the specified port.
+  if (opts.multiTenant && opts.port > 0) {
+    auto pos = cachedServers_.find(opts.port);
```

Comment:
curious, do we need a lock here? Or is multi-thread not a valid use case?

Reply:
You are right, we need one. I was too fixated to the single-threaded bootstrap process with `init_process_group()` and `init_rpc()`, and did not consider other cases. Since this code path is potentially shared by more than one store, we need mutual exclusivity. I will fix it in the next rev.